packages/core/src/indices/vectorStore/VectorStoreIndex.ts

```diff
@@ -21,32 +35,46 @@ import {
 import { BaseNodePostprocessor } from "../BaseNodePostprocessor";
 import { VectorIndexRetriever } from "./VectorIndexRetriever";
 
-export interface VectorIndexOptions {
-  nodes?: BaseNode[];
+interface IndexStructOptions {
   indexStruct?: IndexDict;
   indexId?: string;
+}
+export interface VectorIndexOptions extends IndexStructOptions {
+  nodes?: BaseNode[];
   serviceContext?: ServiceContext;
   storageContext?: StorageContext;
+  imageVectorStore?: VectorStore;
+  vectorStore?: VectorStore;
 }
 
 export interface VectorIndexConstructorProps extends BaseIndexInit<IndexDict> {
-  vectorStore: VectorStore;
+  indexStore: BaseIndexStore;
+  imageVectorStore?: VectorStore;
 }
 
 /**
  * The VectorStoreIndex, an index that stores the nodes only according to their vector embedings.
  */
 export class VectorStoreIndex extends BaseIndex<IndexDict> {
   vectorStore: VectorStore;
+  indexStore: BaseIndexStore;
+  embedModel: BaseEmbedding;
+  imageVectorStore?: VectorStore;
+  imageEmbedModel?: MultiModalEmbedding;
 
   private constructor(init: VectorIndexConstructorProps) {
     super(init);
-    this.vectorStore = init.vectorStore;
+    this.indexStore = init.indexStore;
+    this.vectorStore = init.vectorStore ?? init.storageContext.vectorStore;
+    this.embedModel = init.serviceContext.embedModel;
```

Comment:
@yisding the old code (and the Python code) are inconsistent in retrieving objects from the service context (e.g. `embedModel`). Sometimes, objects are retrieved at construction time, and sometimes, when a function is called.
We should fix this. Options: 
1. Retrieving at construction time, then we can override a setting in the constructor (like in the line above: we can set another vector store instead of using the one from the storage context).
2. Retrieving at function call time, then we can override a setting if needed in the function call 
As some overrides (e.g. the vector store) don't make sense at function call time, I would opt for 1.

Reply:
This is a good point. So I think the vision for service context longer term is for it to go away entirely.

What service context really is is two things: 1. a set of default configurations, and 2. a set of prebuilt objects (for optimization/lack of duplication more than anything).

The problem with it is that, kind of like our classes, it contains state that's not easy to reason with. Often times, the class will have its own state that's somewhat shared with servicecontext, and different classes will have state that's very very similar (and sometimes those classes are also part of servicecontext, which makes things extra fun). And the way we currently do configurations makes sense in some cases, but doesn't make sense in others (for example the temp and prompts are probably things that people want to change more frequently).

That said, there is some necessity to have a set of default configuration. imagine if every time you wanted to use llamaindex to make a single LLM you had to pass back in the url, your key, the system prompt, and everything over again.

So something to think about and to put on our roadmap as to how to handle.